core/src/main/java/org/arquillian/smart/testing/impl/TestStrategyApplierImpl.java

```diff
@@ -35,16 +36,28 @@
     }
 
     public Set<TestSelection> applyOnNames(Iterable<String> testsToRun) {
-        return apply(testsToRun, className -> className);
+        return apply(testsToRun, className -> className, this::selectTestsFromNames);
+    }
+
+    private Collection<TestSelection> selectTestsFromNames(TestExecutionPlanner testExecutionPlanner,
+        Iterable<String> testsToRun) {
+        return testExecutionPlanner.selectTestsFromNames(testsToRun);
     }
 
     public Set<TestSelection> applyOnClasses(Iterable<Class<?>> testsToRun) {
-        return apply(testsToRun, Class::getName);
+        return apply(testsToRun, Class::getName, this::selectTestsFromClasses);
+    }
+
+    private Collection<TestSelection> selectTestsFromClasses(TestExecutionPlanner testExecutionPlanner,
+        Iterable<Class<?>> testsToRun) {
+        return testExecutionPlanner.selectTestsFromClasses(testsToRun);
     }
 
     private <TESTCLASS> Set<TestSelection> apply(Iterable<TESTCLASS> testsToRun,
-        Function<TESTCLASS, String> mapperToName) {
-        final Set<TestSelection> selectedTests = selectTests(configuration);
+        Function<TESTCLASS, String> mapperToName,
+        BiFunction<TestExecutionPlanner, Iterable<TESTCLASS>, Collection<TestSelection>> selectionFunction) {
```

Comment:
I am not sure if using `BiFunction` here is not making it hard to follow (plus `<TESTCLASS>` "generic"). I have a feeling that this is a smell that the code here became overcomplicated (if it wasn't already).

Reply:
We have a mix of high-level, general purpose concepts such as `apply` and then within it `selectTests`. Not sure how we can make it better, but maybe a good start would be to separate `selectTests` logic from applying a strategy (so to call them one after the other instead of one from within another)?

Reply:
Yeah, I do agree that it is overcomplicated - I wanted to avoid copy-pasting and code duplication.
I tried to refactor it, but I'm not sure if I made it better - I introduced a new abstraction - `TestSelector` that has two implementations (based on the class info type - classes/names). It does the whole magic and from the applier you call only `selectTests` or `orderTests`

Reply:
Way nicer :) Let's see if we can further improve it going forward, but this is good enough! Thank you!